### ISSUE TITLE
`Mexpect` for expecting prompt

### ISSUE THREAD

**Issue by talha252** [itublg/calico#1]
Students often ignore the given instructions when they are implementing the assigned algorithm. Sometimes, even though the result they produced is the correct, we failed to evaluate the program properly because they didn't follow the given prompt instructions precisely. Currently, _Calico_ evaluates this situation as **FAILED** but I think it should be evaluated as **ERROR**. To make it clear, an example would be:

```
...
     script:
        - expect: '[Ee]nter [Cc]ards?(.*?):\s+' # timeout: 2
        - send: 'c1'
        - send: 's1'
        - send: 'c1'
        - send: 's1'
        - send: '.'      
        - expect: '[Ee]nter [Mm]oves?(.*?):\s+' # timeout: 2
        - send: 'd'
        - send: 'd'
        - send: 'd'
        - send: 'd'
        - send: 'd'
        - send: '.'
        - expect: '[Ee]nter [Gg]oal(.*?):\s+' # timeout: 2
        - send: '42'
        - expect: '[Ss]core:\s+3' # timeout: 2
        - expect: _EOF_
...
```
Let's assume the student produce the correct result which is `Score: 3`. However, if the previous prompt was `Enter Goal` (without colon) the evaluation would stuck at that point and when the timeout expired, it would procude FAILED. Nevertheless, the result was correct but we didn't even reach that point. Let's change `expect` with `mexpect`.

```
...
     script:
        - mexpect: '[Ee]nter [Cc]ards?(.*?):\s+' # timeout: 2
        - send: 'c1'
        - send: 's1'
        - send: 'c1'
        - send: 's1'
        - send: '.'      
        - mexpect: '[Ee]nter [Mm]oves?(.*?):\s+' # timeout: 2
        - send: 'd'
        - send: 'd'
        - send: 'd'
        - send: 'd'
        - send: 'd'
        - send: '.'
        - mexpect: '[Ee]nter [Gg]oal(.*?):\s+' # timeout: 2
        - send: '42'
        - expect: '[Ss]core:\s+3' # timeout: 2
        - expect: _EOF_
...
```
In this case, if we stuck at `Enter Goal` part, now we expect **ERROR** from _Calico_  rather than **FAILED** . 

In a nutshell, my proposal is that we should use `expect` when we are expecting a **result** from the user and `mexpect` when we are expecting non-critical things like a **menu prompt**.

**Comment by uyaritu** [comment]
If I understand correctly, you're suggesting a "soft" expect which will not fail if the expected data is not received. I would rather not tie it to just prompts and make it more general. It's a nice idea, though I have some problems with ambiguities:

- If Calico reports an error but then receives the correct result on a "hard" expect, how will the case be graded?
- Isn't it possible that an error in the assignment regarding an earlier expect will also consume the correct output that is expected in later stages? Should we just collect all output and check it at the end of the run then?

**Comment by talha252** [comment]
When I was writing this issue, I had another "hard" expect with different output in my mind. So, it would stop the test as soon as an error occured but it would show another output (**ERROR** in this case). However, I liked the _soft expect_ idea. My humble suggestions regarding the ambiguities are:
- If Calico recieves the correct result on a _hard expect_ but an error on a _soft expect_, I think the grade can be given as full point. Because, if we recieve the correct result even with an error, it means there was a little mistake on _soft expect_ (such as a missing colon) which were blocking the test. Nevertheless, to warn the tester for possible errors (so they can debug later if they want to), we may append an asterix character (*) to **PASSED**. 
- It is possible. If Calico reports an error on a _soft expect_ and a failure on _hard expect_, Calico should stop testing and should produce **ERROR** (maybe we can the line number of the _soft expect_ to the report, but not sure). **FAILED** should only be reported if there is a failure on a _hard expect_ and no errors on _soft expect_. With this, the tester understands that there is a problem with a _soft expect_. After the issue in the _soft expect_ is fixed, the test can be rerun as usual. I don't think collecting all outputs and then checking them is a good idea. It will overcomplicate things and it may cause unpredictable results.
